pkg/auditlog/publisher.go

```diff
@@ -0,0 +1,158 @@
+package alog
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// Publisher represents an audit log events publisher that posts events to
+// WorkOS asynchronously.
+type Publisher struct {
+	// The WorkOS api key. It can be found in
+	// https://dashboard.workos.com/api-keys.
+	APIKey string
+
+	// The http.Client that is used to post audit log events to WorkOS. Defaults
+	// to http.Client.
+	Client *http.Client
+
+	// The endpoint used to request Workos. Defaults to
+	// https://api.workos.com/events.
+	Endpoint string
+
+	// The function used to encode in JSON. Defaults to json.Marshal.
+	JSONEncode func(v interface{}) ([]byte, error)
+
+	// The function used to log errors. Defaults to log.Printf.
+	Log func(format string, v ...interface{})
+
+	// The size of the internal queue. Defaults to 512.
+	QueueSize int
+
+	queue  chan Event
+	cancel func()
+	stop   chan struct{}
+	once   sync.Once
+}
+
+// Publish enqueues the given events to be published to WorkOS.
+func (p *Publisher) Publish(events ...Event) {
+	p.once.Do(p.init)
+
+	for _, e := range events {
+		e.Location = defaultLocation(e.Location)
+		e.OccurredAt = defaultTime(e.OccurredAt)
+		p.queue <- e
+	}
+}
+
+func (p *Publisher) init() {
+	if p.Client == nil {
+		p.Client = &http.Client{Timeout: 10 * time.Second}
+	}
+
+	if p.Endpoint == "" {
+		p.Endpoint = "https://api.workos.com/events"
+	}
+
+	if p.JSONEncode == nil {
+		p.JSONEncode = json.Marshal
+	}
+
+	if p.Log == nil {
+		p.Log = log.Printf
+	}
+
+	if p.QueueSize < 1 {
+		p.QueueSize = 512
+	}
+
+	p.queue = make(chan Event, p.QueueSize)
+	p.stop = make(chan struct{})
```

Comment:
Should this just be a `bool`?

Reply:
The struct{} is a struct type with zero elements. It is often used when no information is to be stored. It has the benefit of being 0-sized, so usually no memory is required to store a value of type struct{}